### ISSUE TITLE
Non-prerelease black release

### ISSUE THREAD

**Issue by lig** [psf/black#517]
It looks like Black is stable enough at the moment to release it as `stable` to PyPI. There are plenty of tools and integrations already which rely on `black` package and the fact that it is still in `prerelease` state sometime causes inconveniences, i.e. using it with `pipenv` forces `allow_prereleases` flag to be set to `true`.

**Comment by ambv** [comment]
This will happen soon, you're right. We'll be releasing one last (planned) beta, and then an RC of a stable release.

**Comment by gforcada** [comment]
@ambv not to be pushy/disrespectful, but is that *soon*, actually soon? :sweat_smile:

**Comment by ambv** [comment]
Right. Surely before Christmas, likely not in November though.

**Comment by Aareon** [comment]
Looking forward to this as `pipenv` screams if you try to install without the `--pre` flag

**Comment by Aareon** [comment]
I agree. Being able to streamline Black with other development tools is a huge plus. Black is currently a MUST in all of my projects.

**Comment by lig** [comment]
@ambv Merry Christmas!:)

**Comment by dimaqq** [comment]
Please? 🙏

**Comment by dyve** [comment]
See also 
https://github.com/alanhamlett/pip-update-requirements/issues/23#issuecomment-456661501

This means we cannot update black in `requirements.txt` with pur.

A non-beta version would be much appreciated.

**Comment by Utsav2** [comment]
Black at `HEAD` has a fix for issue 282, and we can't beta test it at Dropbox until there's a release with that issue fixed. Please consider creating a new release soon.

**Comment by Peque** [comment]
@ambv Friendly ping. :innocent:

**Comment by fernandezpablo85** [comment]
@ambv friendlier ping

**Comment by max-wittig** [comment]
@ambv even friendlier ping :P

**Comment by hugovk** [comment]
> Give the last beta a serious spin for the remainder of March! I'll tag the first stable release in early April.

https://twitter.com/llanga/status/1106247630370283523

**Comment by dimaqq** [comment]
Cheers @hugovk , it does feel like formatting got pretty stable in `black`:
2 lines changed in 100KLOC codebase with the latest release :)

**Comment by tribals** [comment]
Any changes on this?

**Comment by zsol** [comment]
I hear @ambv loves making stable releases in April, so I have high hopes ;)

**Comment by ambv** [comment]
FYI, we are sprinting on Black at PyCon now with the target stable release at some point right after.

**Comment by max-wittig** [comment]
What's the status here? It's annoying to use/install a pre-release package :disappointed:

**Comment by Imaclean74** [comment]
Still pending ? is there anything blocking the release creation ? or its a matter of someone's time to actually do the work ?

**Comment by dvf** [comment]
It’s time 🙏

**Comment by fernandezpablo85** [comment]
@ambv ?

**Comment by Juanlu001** [comment]
![Screenshot_2019-07-23 black](https://user-images.githubusercontent.com/316517/61694739-653d5880-ad32-11e9-8c15-54049c24487e.png)

**Comment by ambv** [comment]
Hey everyone, I'm sorry this was slipping so long. Long story short:
- I had to move Black under the psf/ org and this had priority, you can't have a stable release and then change everybody's links next week;
- marking the release as stable is a Big Deal to us because it's when we promise not to fiddle with the style anymore, there's a few last things we need to merge or decide on;
- this project is barely 18 months old, please be a bit patient, it's an unpaid labor of love for the maintainers.

**Comment by lig** [comment]
@ambv I would argue that stable means no style changes in the future though. I'd like to see some sort of style change policy in terms of something like semver. It is obvious, in my opinion, that there always could arise an issue with some formatting. And in case no style change ever after the stable release I would vote for the stable release never appear at all.

**Comment by mrocklin** [comment]
FWIW as a user I would prefer a release before things are fully stable.  Honestly, black is far more stable than most of the software that I use (pandas, scikit-learn, tornado, ...).

I'm also totally fine if you want to fiddle with style in the future.  I would actually prefer that `black` maintainers *not* promise to never fiddle with style.  I'm confident that you all will learn things in the future that you'll want to apply, and I'd prefer that you have some freedom to make those changes.

In general I think that it's perfectly acceptable to release software before it is fully stable.  The world runs on and benefits from imperfect software.  Obviously y'all should do whatever you think is best, but I think that there is a lot of good that black can do today in its current form.

**Comment by rpdelaney** [comment]
@mrocklin is right.  "We moved to psf" and "we aren't getting paid" are excellent reasons to delay a stable release. But I think "we need to settle some style issues permanently first" isn't as compelling.

A commitment to _never_ making style changes again is likely to lead you to situations where you have to either compromise, or leave in sub-optimal behavior in corner cases that haven't been considered yet.

I'm using black with a pre-commit hook and installing it as a dev dependency via pipenv. That means I can pin the black version to the project if I want to prevent future style changes mucking up the git blame and whatnot. So style changes are not a threat to me.

But it is a threat, however small, that installing black with pipenv is a bit annoying since it requires the use of the `--pre` switch. As long as there is no stable release, any black version I pin to the project will require me to document the use of the `--pre` switch for other developers now and forever. 

Besides, "we won't make style changes" is a policy decision, not a technical limitation, meaning it could always be reversed (and so anyone who really truly cares about this will have to pin the version somehow anyway).

Stable release please. 🙏

**Comment by thomasf** [comment]
Why does a stable release matter to you if it in fact does not have a stable output?

I fully support the pre-release until formatting is decided MO and preferably moving to semantic versioning so that 1.x never will break formatting or usage.

**Comment by Kilo59** [comment]
> I'm using black with a pre-commit hook and installing it as a dev dependency via pipenv. That means I can pin the black version to the project if I want to prevent future style changes mucking up the git blame and whatnot. So style changes are not a threat to me.
> 
> But it is a threat, however small, that installing black with pipenv is a bit annoying since it requires the use of the `--pre` switch. As long as there is no stable release, any black version I pin to the project will require me to document the use of the `--pre` switch for other developers now and forever.
 
@rpdelaney 
If you pin the version you don't need to use pipenv `--pre` argument.
Just define set it in your Pipfile as `black = "==19.3b0"`

**Comment by max-wittig** [comment]
@thomasf it doesn't help that pipenv classifies the whole project as preview, if you're using the `--pre` argument

**Comment by rpdelaney** [comment]
@max-wittig Based on his comment I assume Thomas will rate that as a strike against pipenv, not black.

**Comment by dimaqq** [comment]
I feel that this thread touches on a related issue — the release cadence could be better.
About 50 PRs were merged since last release.
I think I'll start ti pin `black` to a commit hash rather than release soon!

Wrt. having or not having a stable release, there are two ways to look at it:
* ideology "what is stable?", and
* usability issue "silly, but users come first"

**Comment by dvf** [comment]
It seems like we're dealing with two kinds of versions here: one is the actual version of Black (the code itself), the other is the rules of how Black formats code. I think it would make sense to separate these. 

I'd love to hear your thoughts @ambv

**Comment by paugier** [comment]
We also would need a stable (actually not beta) release of Black.

Black is used in [Transonic](https://transonic.readthedocs.io), a new "meta" Python compiler for scientific/numerical codes. In Transonic, we produce Python code from AST analyses and Black is used to have nice and readable codes for the users/developers (and also to avoid useless compilations after style changes). Small style changes between Black versions are really not an issue for us.

I don't personnaly use Pipenv, but some projects using Transonic use Pipenv in their CI. If we add Black as a strong dependency for Transonic (as we should do), all projects using Transonic become incompatible with tox-pipenv.

I don't think it would make sense to pin the Black version for Transonic (which would lead to other problems).

I understand that you care about the stability of Black's results, but (i) semver would be fine to express that and (ii) the idea of two version codes (one for Black's code and one for Black's result) seems good.

**Comment by rpdelaney** [comment]
> I don't think it would make sense to pin the Black version for Transonic (which would lead to other problems).

I agree -- that's evident from your relaxed approach toward style changes within black. A project that has decided to take a more conservative approach still has this problem, though.

> the idea of two version codes (one for Black's code and one for Black's result) seems good.

This is an interesting idea but I don't know how a project could pin to an output version, or how conflicts would be resolved when the app version and output version are incompatible, etc etc. As far as I'm aware none of the existing utilities (setuptools, pipenv, poetry etc) would support this.

**Comment by paugier** [comment]
Dear @thomasf and @carver,

You reacted with thumbs down emoji to my recent message. I guess it means that you think there is a problem with this message. Can you please tell me why?

I'm wondering if it is really a good idea to use this negative emoji for such polite and neutral message... i just express the need of a user.

**Comment by thomasf** [comment]
No it just means that I don't agree and your point is already present further up in the comments so just like this comment which I had preferred not to write it just makes reading the issue more of a chore for anyone new or a maintainer to read through.    👎 ( - 1 ) is the most neutral way to just show disagreement by vote without further making the whole comment thread more noisy.

**Comment by mar10** [comment]
I think (as many others here in the thread) that a real/stable release would make our lives easier.

If we are concerned about finding out later that we need/want changes in the formatting behavior this can be expressed with SemVer. I don't think it's feasible nor necessary to have to different version numbers for the API and the formatting.  
Couldn't we simply bump the **major** version every time when a breaking CLI change is introduced and every time a relevant formatting change is implemented?

I'd call a formatting change *relevant* if it has the potential to mess up your code 'blame' in a relevant way (which is also a sign that it could make merging difficult).

This would give the maintainers the freedom to change behavior later without being tied to an earlier spec.

And nobody will be surprised by implicitly updating black, but every one could enjoy the new coolness by deliberately incrementing black's requirement.

**Comment by paugier** [comment]
> Judging from the other comments in this thread you still have the option to just pin the pre release version to get around that limitation/issue with pipenv if you don't want to wait for a non pre release version which isn't a huge hassle in any way, it even protects your project from sudden style changes.

It seems that you didn't read the message that you disagreed :slightly_smiling_face: 

My project uses black at run time and we don't care at all about "sudden style changes".

Pinning the Black version for projects using Black at run time is really not a good idea. if different projects pin different Black versions, they become incompatible just because of Black.

So unfortunately, I don't think there is with only beta releases a way to express the fact that we need Black without potential issues for downstream projects (tox-pipenv broken or potential  incompatibility between projects that pin black version).

i don't say that Black maintainers should release a stable version (I don't know). I say that for projects using Black at runtime, no "stable" version is "a huge hassle" and a "stable" version (which could be 0.0.n to clearly show that Black is not "stable") would help a lot. 

That said, I really understand that it is not Black's priority to help unimportant users if it is for the good of many other users.  But please don't tell me that we are silly to care about a stable version and that it is so easy to deal with this problem (or give me a proper solution). I start to think the proper solution today is just to use another tool than Black. It's just inconvenient, because current versions of Black do very well the job.

**Comment by pcraciunoiu** [comment]
Been using black for a long time, and I love it.

I'd like to see a non-prerelease as well. I have to use `allow_prereleases = true` in `Pipfile` because of this project, and it causes other projects that are more critical to functionality to pull in prereleases as well.

I'm not saying there's no workaround, but if a project has been out for years now... can it have at least one release please?

One workaround is to pin releases for other dependencies. But one point of the lock file with `pipenv` is to keep updating your dependencies. That would break a lot less frequently for me if this project had a release.

**Comment by hnykda** [comment]
Hi. Should we give up on black ever have stable release?

**Comment by ambv** [comment]
https://twitter.com/llanga/status/1187675966694449152

**Comment by hnykda** [comment]
Thanks for your hard work, black is awesome! It may be a language barrier, but I don't understand the tweet ending: 

> We *will* have a stable Black with 19 in the front!

Considering that there has been already about a year since the first "for sure" and then once again after Pycon, I am curious if there is any reason to believe that it's going to be different this time? :grin:

**Comment by hugovk** [comment]
Black uses [calver](https://calver.org/). It means there will be a stable release in 2019 :)

**Comment by max-sixty** [comment]
Thanks for the recent release! We had some small breaks over at https://github.com/pydata/xarray/pull/3456 and it made me think of an approach that I've seen work well with other similar projects: 

_When introducing breaking changes, deprecate the previous code style with warnings, and only introduce them as errors in a later version_

That way, there's no 'big switch' and CI doesn't break without pinning. Rather, projects can upgrade their syntax whenever works for them. On the downside, running `black --check .` doesn't 100% mirror whether changes would happen with `black .`.

Thanks again for the awesome tool!

**Comment by ambv** [comment]
> On the downside, running `black --check .` doesn't 100% mirror whether changes would happen with `black .`

I'm sorry, are you sure? `--check` is doing exactly what the regular run is doing, it's just not writing the file back to disk. Please ensure you don't have any integration magic like using the wrong configuration in the `--check` run. If you find an example where the behavior is indeed different, it's a rather severe bug so I'd like to know about it.

**Comment by max-sixty** [comment]
> `--check` is doing exactly what the regular run is doing

Looks like my post wasn't clear! I was listing this as a downside to a proposal, not current behavior.

i.e. if breaking changes were initially deprecated, then `black --check .` would pass with warnings, but `black .` would fix code to the new format.

**Comment by max-wittig** [comment]
Still no stable release with 19 in front as promised here: https://twitter.com/llanga/status/1187675968007278592 :(
Can we have one with 20 in front soon?

What's blocking this? How can the community help?

**Comment by revmischa** [comment]
By having to enable pre-releases in my Pipfile I've started getting beta versions of libraries and breaking my stuff. Black is the only dependency I have that is a pre-release. Please release a stable version.

**Comment by larsrinn** [comment]
Huge upvote on this. I don't see a reason for not dropping the `b0`. We've been using black for 1.5 years on a not so small project and it didn't cause any issues, so it seems like it's pretty stable to me. 

Of course there are some changes here and there after version upgrades, but I don't think this is going to go away ever. For exactly this reason, we pin black to a specific version, so we get clean diffs for PRs that upgrade black (and don't change anything else). Maybe this could be the recommended behaviour?

**Comment by hnykda** [comment]
I am really greateful for all the work devs are putting into this great (opens source!) lib that Python ecosystem is gaining so much from.

But I also have to admit that I am quite disappointed with the communication related to this issue. It has been more than a year since the announcement of the non-prerelease and then several more wrong estimates, without really explaining why or what is happening. I take this quite seriously, as it cascades to suboptimal decision on people who rely on this. E.g. we would have forked black a half year ago already so we wouldn't have to pollute our Pipenv envs with prereleases... No offence please, I am just trying to share why it's annoying in the case someone doesn't realize this.

**Comment by oteph** [comment]
> Been using black for a long time, and I love it.
> 
> I'd like to see a non-prerelease as well. I have to use `allow_prereleases = true` in `Pipfile` because of this project, and it causes other projects that are more critical to functionality to pull in prereleases as well.
> 
> I'm not saying there's no workaround, but if a project has been out for years now... can it have at least one release please?
> 
> One workaround is to pin releases for other dependencies. But one point of the lock file with `pipenv` is to keep updating your dependencies. That would break a lot less frequently for me if this project had a release.

I just noticed why i was getting beta build for various libs that kept breaking stuff. Turns out someone had added pre_release so black could be installed with pipenv.

The Solution for me is to remove the flag and to pin the black version. 
```toml
[dev-packages]
black = "==19.10b0"
```
or

`pipenv install --dev black==19.10b0
`

With this there is no need to set the pre release option in pipfile. I might come back in a year to check if the maintainers managed to stop abusing the pre release flag.

**Comment by carver** [comment]
> For exactly this reason, we pin black to a specific version, so we get clean diffs for PRs that upgrade black (and don't change anything else). Maybe this could be the recommended behaviour?

I disagree. I still think semantic versioning is a worthwhile goal (vs pinning to avoid breaking changes). There are a number of projects that are waiting to use `black` until it gets locked stable, including some of mine.

Of course, we'd love to have the stable release out as soon as possible, but it's *more* important to us to have a stable release that's committed to semantic versioning.

**Comment by revmischa** [comment]
Okay so release 1.0.0 and if you need to make incompatible changes then release 2.0.0?

**Comment by Peque** [comment]
Black 1.0.0: The uncompromising and aerospace-grade Python code formatter. :smile:

**Comment by revmischa** [comment]
> > Been using black for a long time, and I love it.
> > I'd like to see a non-prerelease as well. I have to use `allow_prereleases = true` in `Pipfile` because of this project, and it causes other projects that are more critical to functionality to pull in prereleases as well.
> > I'm not saying there's no workaround, but if a project has been out for years now... can it have at least one release please?
> > One workaround is to pin releases for other dependencies. But one point of the lock file with `pipenv` is to keep updating your dependencies. That would break a lot less frequently for me if this project had a release.
> 
> I just noticed why i was getting beta build for various libs that kept breaking stuff. Turns out someone had added pre_release so black could be installed with pipenv.
> 
> The Solution for me is to remove the flag and to pin the black version.
> 
> ```toml
> [dev-packages]
> black = "==19.10b0"
> ```
> 
> or
> 
> `pipenv install --dev black==19.10b0 `
> 
> With this there is no need to set the pre release option in pipfile. I might come back in a year to check if the maintainers managed to stop abusing the pre release flag.

This doesn't work for me at all.

```
[pipenv.exceptions.ResolutionFailure]:       pipenv.exceptions.ResolutionFailure: ERROR: ERROR: Could not find a version that matches black==19.10b0,>=19.3b0
[pipenv.exceptions.ResolutionFailure]:       Skipped pre-versions: 18.3a0, 18.3a0, 18.3a1, 18.3a1, 18.3a2, 18.3a2, 18.3a3, 18.3a3, 18.3a4, 18.3a4, 18.4a0, 18.4a0, 18.4a1, 18.4a1, 18.4a2, 18.4a2, 18.4a3, 18.4a3, 18.4a4, 18.4a4, 18.5b0, 18.5b0, 18.5b1, 18.5b1, 18.6b0, 18.6b0, 18.6b1, 18.6b1, 18.6b2, 18.6b2, 18.6b3, 18.6b3, 18.6b4, 18.6b4, 18.9b0, 18.9b0, 19.3b0, 19.3b0, 19.10b0, 19.10b0
[pipenv.exceptions.ResolutionFailure]: Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.
  First try clearing your dependency cache with $ pipenv lock --clear, then try the original command again.
 Alternatively, you can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.
  Hint: try $ pipenv lock --pre if it is a pre-release dependency.
```
And if you install with `--pre` it adds `allow_prereleases = true` to your Pipfile.

**Comment by kierun** [comment]
Note that using [Poetry](https://python-poetry.org/), you can have:

```
[tool.poetry.dev-dependencies]
black = "^19.3b0"
```

However, this does not help the `pipenv` folks at all…

**Comment by dimaqq** [comment]
Since this discussion started my team move `black` into precommit and we don’t need it in deps at all :)
That’s probably the main strength of black, that it’s so reliable (as in doesn’t break code) that we’re using it as invisible part of the flow “Also reformat all modified py files” rather than a test “is this code well-formatted?”

**Comment by oteph** [comment]
> > > Been using black for a long time, and I love it.
> > > I'd like to see a non-prerelease as well. I have to use `allow_prereleases = true` in `Pipfile` because of this project, and it causes other projects that are more critical to functionality to pull in prereleases as well.
> > > I'm not saying there's no workaround, but if a project has been out for years now... can it have at least one release please?
> > > One workaround is to pin releases for other dependencies. But one point of the lock file with `pipenv` is to keep updating your dependencies. That would break a lot less frequently for me if this project had a release.
> > 
> > 
> > I just noticed why i was getting beta build for various libs that kept breaking stuff. Turns out someone had added pre_release so black could be installed with pipenv.
> > The Solution for me is to remove the flag and to pin the black version.
> > ```toml
> > [dev-packages]
> > black = "==19.10b0"
> > ```
> > 
> > 
> > or
> > `pipenv install --dev black==19.10b0 `
> > With this there is no need to set the pre release option in pipfile. I might come back in a year to check if the maintainers managed to stop abusing the pre release flag.
> 
> This doesn't work for me at all.
> 
> ```
> [pipenv.exceptions.ResolutionFailure]:       pipenv.exceptions.ResolutionFailure: ERROR: ERROR: Could not find a version that matches black==19.10b0,>=19.3b0
> [pipenv.exceptions.ResolutionFailure]:       Skipped pre-versions: 18.3a0, 18.3a0, 18.3a1, 18.3a1, 18.3a2, 18.3a2, 18.3a3, 18.3a3, 18.3a4, 18.3a4, 18.4a0, 18.4a0, 18.4a1, 18.4a1, 18.4a2, 18.4a2, 18.4a3, 18.4a3, 18.4a4, 18.4a4, 18.5b0, 18.5b0, 18.5b1, 18.5b1, 18.6b0, 18.6b0, 18.6b1, 18.6b1, 18.6b2, 18.6b2, 18.6b3, 18.6b3, 18.6b4, 18.6b4, 18.9b0, 18.9b0, 19.3b0, 19.3b0, 19.10b0, 19.10b0
> [pipenv.exceptions.ResolutionFailure]: Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.
>   First try clearing your dependency cache with $ pipenv lock --clear, then try the original command again.
>  Alternatively, you can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.
>   Hint: try $ pipenv lock --pre if it is a pre-release dependency.
> ```
> 
> And if you install with `--pre` it adds `allow_prereleases = true` to your Pipfile.

```
oteph@oteph-pc:~$ mkdir black_test
oteph@oteph-pc:~$ cd black_test/
oteph@oteph-pc:~/black_test$ pipenv install --dev black==19.10b0
Creating a virtualenv for this project…
Pipfile: /home/oteph/black_test/Pipfile
Using /usr/bin/python3 (3.6.8) to create virtualenv…
⠧ Creating virtual environment...Already using interpreter /usr/bin/python3
Using base prefix '/usr'
New python executable in /home/oteph/.local/share/virtualenvs/black_test-3N6-Y2Bp/bin/python3
Also creating executable in /home/oteph/.local/share/virtualenvs/black_test-3N6-Y2Bp/bin/python
Installing setuptools, pip, wheel...
done.

✔ Successfully created virtual environment! 
Virtualenv location: /home/oteph/.local/share/virtualenvs/black_test-3N6-Y2Bp
Creating a Pipfile for this project…
Installing black==19.10b0…
Adding black to Pipfile's [dev-packages]…
✔ Installation Succeeded 
Pipfile.lock not found, creating…
Locking [dev-packages] dependencies…
✔ Success! 
Locking [packages] dependencies…
Updated Pipfile.lock (74ab12)!
Installing dependencies from Pipfile.lock (74ab12)…
  🐍   ▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉ 8/8 — 00:00:01
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.
oteph@oteph-pc:~/black_test$ cat Pipfile
[[source]]
name = "pypi"
url = "https://pypi.org/simple"
verify_ssl = true

[dev-packages]
black = "==19.10b0"

[packages]

[requires]
python_version = "3.6"
oteph@oteph-pc:~/black_test$ 
```

seems to work fine

**Comment by ambv** [comment]
Hey all,
I'd just like to say I'm sorry for giving a few release dates and then not delivering on this front. This is due to two things. First, I have some probably unreasonable wants for the stable version here. Before I meet them, I would not feel good tagging the release. Second, this is not a paid gig for me and as such it sadly falls into lower priority.

I won't be making a mistake of promising a date anymore but I can tell you this much:
- next week I'm going on a week-long vacation to Malta, which should ideally bring some much needed sunshine into my miserable Polish winter existence;
- after that I will be working hard with the team to finalize the release. I will create a "stable" milestone on GitHub and link issues I consider crucial for inclusion. You'll be able to see how we're doing through that.

Again, sorry for the disappointment. We'll get out of that ditch together.

**Comment by mrocklin** [comment]
We love you and all you do @ambv !  Enjoy the sun.

**Comment by hbmartin** [comment]
You're awesome @ambv ! Are there any labels for issues that a newbie to the project can take on?

**Comment by lig** [comment]
@ambv I'm deeply sorry but there is no `stable` label on any issue at the moment.

**Comment by elreydetoda** [comment]
> Hey all,
> I'd just like to say I'm sorry for giving a few release dates and then not delivering on this front. This is due to two things. First, I have some probably unreasonable wants for the stable version here. Before I meet them, I would not feel good tagging the release. Second, this is not a paid gig for me and as such it sadly falls into lower priority.
> 
> I won't be making a mistake of promising a date anymore but I can tell you this much:
> 
> * next week I'm going on a week-long vacation to Malta, which should ideally bring some much needed sunshine into my miserable Polish winter existence;
> * after that I will be working hard with the team to finalize the release. I will create a "stable" milestone on GitHub and link issues I consider crucial for inclusion. You'll be able to see how we're doing through that.
> 
> Again, sorry for the disappointment. We'll get out of that ditch together.

@ambv, would you consider adding a sponsor option? So that way it helps you out and shows how much people appreciate your work in another way than just words. While I am sure it won't pay as much as your current job (probably development, but not going to assume :slightly_frowning_face: ) at least it will be something to help offset your time working on this project. Which thank you for all your continued work over these past two years! I know sticking with projects and tying your name to something is hard! :heart:

**Comment by johnthagen** [comment]
One way to mitigate the perception risks with a stable release is to specifically instruct people to **pin** their version of `black` used in their project.

Prettier is a very popular code formatter for JavaScript/TypeScript and [they do just that](https://prettier.io/docs/en/install.html):

> `npm install --save-dev --save-exact prettier`

This pins the exact version of Prettier in the project's `package.json`

Later in their docs they explain this clearly:

>...because we change how code is formatted in each release! It’s important to have a locked down version of Prettier in your package.json

I think `black` could take a similar stance and be clear up front that you should pin exact versions, and give examples on how to do this in `poetry`, `pip-tools`, `pipenv`, etc. This then gives `black` the needed flexibility to evolve and improve without breaking user's projects who are pinning properly.

**Comment by ichard26** [comment]
If we ever go with that route, we should also recommend folks to lock in their Black configuration in a `pyproject.toml` file in the case we ever change a default (seems unlikely, but eh).

**Comment by mUtterberg** [comment]
> One way to mitigate the perception risks with a stable release is to specifically instruct people to **pin** their version of `black` used in their project.
This is exactly what brought me here. I tried to lock the version of `black` that I am using with `pipenv`, but was only able to do so by allowing pre-releases for my entire virtual environment. Would love to be able to pin a 1.0+ and remove the pre-release allowance from my Pipfile.

**Comment by Kilo59** [comment]
@mUtterberg

https://github.com/psf/black/issues/517#issuecomment-573125735

**Comment by lig** [comment]
@Kilo59 the problem is that it is not uncommon to have another package with available pre-release in the same Pipfile which is not necessarily necessary to allow installing.

**Comment by ambv** [comment]
The next episode in this saga has just been released. 20.8b0 is hopefully the last beta. There's many things I wanted to fix before claiming Black is stable but the minimal we need to get in is:
- fix for the currently known issues with formatting instability
- fix stability of `--experimental-string-processing`

If we can't get the latter stable enough we'll just remove it. But the former is very important to fix.

**Comment by ambv** [comment]
That first issue is tracked here: https://github.com/psf/black/issues/1629

**Comment by intgr** [comment]
Sorry if this is the wrong forum, but I just wanted to say 🌟 **big thank you** 🌟 for the improvements in this release! My experience with Black 19.10b0 was rocky, but 20.8b1 has been perfect so far.

**Comment by cooperlees** [comment]
Thanks - It's nice to hear this every now and then outside of the usual "why did you change X" and "black should do y" issues. As someone who helps out with black, I appreciate it.

**Comment by ichard26** [comment]
Glad to hear you've had a great experience with Black! *\*god I really do sound like a salesman 😅*\* The common "thanks" in bug reports, feature requests, or design suggestions is always nice... but yeah hearing a "Thank you!" without strings attached definitely lifts up my mood a bit as maintainers[1] typically don't hear it often. 

[1]. Technically I am \*not\* a maintainer but a triager, but I like to think I also help out with Black 😉 

Thanks, but let's also not forget about the [many outside contributors to Black](https://github.com/psf/black#authors) 😃

**Comment by elreydetoda** [comment]
> > Hey all,
> > I'd just like to say I'm sorry for giving a few release dates and then not delivering on this front. This is due to two things. First, I have some probably unreasonable wants for the stable version here. Before I meet them, I would not feel good tagging the release. Second, this is not a paid gig for me and as such it sadly falls into lower priority.
> > I won't be making a mistake of promising a date anymore but I can tell you this much:
> > 
> > * next week I'm going on a week-long vacation to Malta, which should ideally bring some much needed sunshine into my miserable Polish winter existence;
> > * after that I will be working hard with the team to finalize the release. I will create a "stable" milestone on GitHub and link issues I consider crucial for inclusion. You'll be able to see how we're doing through that.
> > 
> > Again, sorry for the disappointment. We'll get out of that ditch together.
> 
> @ambv, would you consider adding a sponsor option? So that way it helps you out and shows how much people appreciate your work in another way than just words. While I am sure it won't pay as much as your current job (probably development, but not going to assume slightly_smiling_face ) at least it will be something to help offset your time working on this project. Which thank you for all your continued work over these past two years! I know sticking with projects and tying your name to something is hard! heart

Hey @ambv, that is awesome news!!! Thank you for all of your and the teams hard work!!

I was curious though, any chance I will get a response to my question above :arrow_up: ? I want to try and help with the only way I have time to do, and that is to try and help sponsor the project. I know it won't be very much, but something like a patreon with different tiers of support so that people whom can't help by coding (either because of time or expertise) can support this awesome project in a different way than code?

**Comment by happyteque** [comment]
PLEASE CAN YOU JUST RELEASE THE FLIPPING TOOL??!? 

It will never be perfect. But it's good enough for every user. It's gone WAY beyond a joke now. 

Thanks

**Comment by ambv** [comment]
> PLEASE CAN YOU JUST RELEASE THE FLIPPING TOOL??!?

Did you expect this sort of message to be effective? You're a random anonymous stranger, coming to my issue tracker to yell at me because you don't like how a free tool's development is managed.

My advice, and this is coming straight from the heart, is: go away.

**Comment by happyteque** [comment]
Obviously NOTHING is going to be effective, since you appear to be a very stubborn man. At least I have conveyed some of the frustration many developers feel about it.

> You're a random anonymous stranger, coming to my issue tracker to yell at me because you don't like how a free tool's development is managed.

Partly correct - besides the shouting I also came here to offer some advice: please release the tool, don't be a perfectionist. We are all using it in production anyway, you are not preventing any pain, but causing it.

**Comment by thomasf** [comment]
> We are all using it in production anyway, you are not preventing any pain, but causing it.

Not me, I won't add it to any lage project until it's released. An external example: Django will adopt black but not until a release version that won't change formatting is available.

If it's so important to you, just clone the source, change the version number and upload it to pypi under a different name and you are done.

**Comment by ambv** [comment]
> I also came here to offer some advice

Oh please. Your "advice" uses all caps, excessive numbers of exclamation marks and question marks, calls our struggle a joke, *and* is factually incorrect (the tool you're claiming is "good for every user" has 315 open issues at the time I'm writing this comment).

Quite a feat for such a short comment from an anonymous person who offered no help beyond unsolicited "advice". Continuing this conversation is in fact the opposite of help, escalating negative emotions and noise on an already busy issue tracker.

**Comment by max-sixty** [comment]
Without wanting to add a comment telling the maintainers what to do  — could I suggest the spammy comments are marked as spam, rather than engaged with?

**Comment by timothycrosley** [comment]
There are many benefits to being careful and methodical when building out a tool as ambitious as Black. As just one example, this care has enabled other commonly used libraries in the Python linting and formatting space to ensure they are fully compatible with Black. Black's conservative release approach has given projects time to do this proactively instead of retroactively. I aspire to the level of discipline and care that @ambv and others who work on this project have taken. I don't follow many projects, but I follow Black as it is a constant source of inspiration for how I can better maintain and contribute to OSS projects.

**Comment by JelleZijlstra** [comment]
Answering @elreydetoda's question from above: I think the best way to help get Black to a stable state really is to get bugs fixed, either by writing the code yourself or by helping get the word out so people who are interested in writing code can find us.

I'd recommend starting with the issues in [unstable formatting](https://github.com/psf/black/issues?q=is%3Aissue+is%3Aopen+label%3A%22unstable+formatting%22) and [invalid code](https://github.com/psf/black/labels/invalid%20code), because those tend to be more straightforward (there's a clear repro case, and clearly wrong behavior). Ideally, stable Black won't have any known crashes. As a maintainer, I'd love to review PRs fixing all of these crashes.

**Comment by ankostis** [comment]
(my apologies for being too late in the party)
Maybe the issue with `pipenv` wouldn't exist if `black`'s version was [PEP440](https://www.python.org/dev/peps/pep-0440/) compatible, with 3 parts (major, minor, build),
AND the major was 0.
If i remember correctly, pip installs pre (non-stable) dependencies that have versions like `0.1.2`,
which makes sense: `black` is at exactly this spot, without having any stable release since its inception,
but still being worked on but many projects downstream depend on it.

So in a nutshell, would you consider to switch to a PEP440 version, with the next release using a new "epoch" like this: `2!0.20.8b2`?

**Comment by astrojuanlu** [comment]
@ambv has stated his very clear conditions to make a stable black release. Many of us subscribed to this issue to hear updates from the Black development team about progress on those conditions. @JelleZijlstra has stated above "the best way to help get Black to a stable state really is to get bugs fixed, either by writing the code yourself or by helping get the word out so people who are interested in writing code can find us". Please keep this thread informative and, if you want to help, make yourself useful in the ways the developers have kindly encouraged us all to do so.

**Comment by henryiii** [comment]
@ambv If it makes you feel any better, NumPy has 1.9k issues and is good for lots of users. :) The number of issues is more indicative of the level of use than it is of the quality of the library. Bad bugs (like #1629) are what matter.

Also, clang-format changes sometimes between releases and is _much_ harder to pin than a Python library. ;)

**Comment by hadialqattan** [comment]
Please see: #1746

**Comment by CoburnJoe** [comment]
I too have just been bitten by allowing preleases with Pipenv specifically for Black...
There seem to be numerous ideas around here from "fix all the bugs and release it" to "release it with a major version 0".
I don't see why releasing as a major version has to mean "all the bugs are fixed".

**Comment by elreydetoda** [comment]
@CoburnJoe, you don't have to enable pre-releases to install black with pipenv, you can do what @massenz suggested and just pin the black version ( https://github.com/psf/black/issues/517#issuecomment-449677652 ).

While it isn't ideal, because that means you will have to update it manually, it allows you to respect @ambv's wishes of waiting ( since it is his project ) and also have a stable environment with black in it.

I have a function that sets up all the linters ( I either want to use or plan to use ), and it just curls to get the most recent black version from github so it can pin it ( as a dev dep w/Pipenv ).

If you want though, you can just use these 3 commands ( in an alias or function if you want ) to pin it to the most recent version:
```bash
black_latest_release="$(curl -sLI -o /dev/null -w '%{url_effective}' "https://github.com/psf/black/releases/latest")"
black_latest_release_tag="${black_latest_release##*/}"
# remove --dev if you want it normally
pipenv install --dev "black==${black_latest_release_tag}"
```

**Comment by av1m** [comment]
> Please see: #1746

I think that this issue is not blocking to obtain the status of `stable`.
Having the entire code written to a single file can be problematic but is another topic (I think).
The point is that black is stable today. I use it in several projects and I am very satisfied with the result.

Certainly a big refactoring is necessary but the code is still segmented in `__init__.py`.
So a first step could be to split the classes into files, then categorize the classes by topic etc.

For information purpose, pycharm gives me this error in `requirements.txt`, but installs the dependency fine.
```
No matching package version found: 'black ~ = 20.8b1'
```

**Comment by ichard26** [comment]
Yeah, please note that issue mention was not by a collaborator of this project and wasn't confirmed by a maintainer so it certainly is not blocking this issue. I also agree that it shouldn't block this, while yes it's important, it's not a showstopper issue (it's not even visible to end users).

**Comment by hnykda** [comment]
Note that https://github.com/psf/black/issues/1746 got in `main` anyway :grin:

**Comment by astrojuanlu** [comment]
There are no "unstable formatting" or "invalid code" issues left, however there's a bunch of them tagged "stable" https://github.com/psf/black/labels/stable. From https://github.com/psf/black/issues/517#issuecomment-578135675

> I will create a "stable" milestone on GitHub and link issues I consider crucial for inclusion.

A bunch of those are labeled as "bug" as well, so I suppose that people willing to see this stable release happening can send pull requests fixing them (I'm not a maintainer though, so don't take this as official)

**Comment by massenz** [comment]
😆 this issue has been around for almost three years now, so I guess it's safe to assume we'll *never* have a "stable" `black` release.
And this, really, doesn't matter: The tool is awesome as it stands, and incredibly helpful, so I guess we can all live with it, and make pace with having to work around the "pre-release" versioning.

I find it hilarious that the issue lived longer than `pipenv` (RIP) did 😆 
Every other build/release tool in the Python universe doesn't care about pre- or release versioning, so I think all is well with the universe.

Again, I [cannot praise highly enough](https://codetrips.com/2018/12/24/uncompromising-python-formatting/) the developers of `black` - thank you folks for sharing the awesomeness.

**Comment by jamesmyatt** [comment]
Is pipenv dead? I thought it was resurrected last year.

**Comment by ichard26** [comment]
Hello everyone :wave:,

First of all I want to make it clear that I'm not announcing a stable release or a date or really any concrete plans. We've been burned too many times by suggesting dates and the whatnot. What I do want to do is bring more transparency and let y'all know what the project has been up to since the last major update (which yikes is from 10 months ago!).

### Project governance and maintainership

First of all, everyone's favourite, the things needed to support an OSS community. I got a quite a few updates here:

- Since April I am now a full blown maintainer so you'll be seeing me merge PRs on top of whatever I used to do :)
- We got a new collaborator quite literally yesterday: @felix-hilden :tada:, they've been helping out all around the project from documentation, bug fixes, to project policy. I can say on the behalf of the core team that we are incredibly happy having them on our team and thankful for all they have done. 
- We've partnered with Python Discord! Since our Freenode IRC channel (which wasn't really used TBH) was no longer a good option we had to look for a new home for our real-time communication. Python Discord (**opinion:** which by the way is an awesome server) is that new home. You'll find the core team and other members of the community hanging out and talking in the `#black-formatter` text channel. [Here's an invite](https://discord.gg/RtVdv86PrH).

### Releasing

Okay, ok, this deserves its own little section of my update. So, you have probably noticed we have been very busy on the releasing front lately (we've made 6 releases in the span of roughly two months!). All thanks go to @cooperlees for stepping up and not only encouraging the rest of us to release but also offering to conduct them (leading the creation of the BRMFN role, meaning _Black. Release. Manager. For. Now_). 

Nowadays we have strong release automation that handle PyPI, Docker, and OS-native binaries, and [releasing documentation](https://black.readthedocs.io/en/latest/contributing/release_process.html) to make it require far less initial work to get the ball rolling on a release. Due to all of these changes, we are aiming to release much more frequently then we used to. This gets new features and bug fixes to end users faster and reduces the impact (and therefore riskiness) of one single release. 

For more technical changes, please see the changelog: https://black.readthedocs.io/en/latest/change_log.html.

### Other noteworthy changes

- The project has been cleaned up significantly. No longer are the days where our monolithic `src/black/__init__.py` file reigned supreme. The documentation has also seen a similar overhaul. Credit goes to @ambv and @ichard26 (and @felix-hilden too) respectively. This should make future changes way less painful and just make it easier to contribute.
- Our issue tracker is way better shape now. While we haven't been able to push the open issue number that low, we have been able to undo the effects of 2020 :p There has been also an overhaul of the labels. Now knowing the status of the issue tracker is way easier and the chaoticness is very so slightly more organized :) Credits go to @JelleZijlstra and me. 

### So what's next?

We will continue working on getting Black closer to a state where we would be happy declaring Black (and drum roll please...) stable. Currently we're trying to improve the stability of Black's new string processing abilities [so we can turn it on by default](https://black.readthedocs.io/en/latest/the_black_code_style/future_style.html#improved-string-processing). Personally I'll be working on continuing the documentation overhaul effort, writing news docs and updating old ones. I've also picked up a new project on finally getting Black compiled with mypyc for performance improvements. No promises on when (or even if) I'll get them done, but please know we are making progress.

In terms of future releases, we'll pushing a new release soon in preparation for another release (I said we're moving quickly with releasing nowadays) where the hopefully non-experimental string processing is turned on by default.

Looking further into the future, we will unfortunately rewrite pretty much the whole project to achieve Python 3.10+ support since our parser is straight up incompatible. More details [tracked in this issue](https://github.com/psf/black/issues/2318).

### Get involved

- First of all, we're looking for feedback for a few things. The only main requirement is that you use Black and can be  constructive :)
  - Right we are considering dropping Python 2 only syntax support (so like for example Black will stop accepting the old backtick repr syntax as valid input; also we will probably keep the compatibility layers that avoids introducing new syntax) and more generally all EOL'd versions. Send your feedback to https://github.com/psf/black/issues/2251.
  - And if you've contributed to this project (sidenote: thank you!) I am looking for feedback on the contributing experience here. We can always do better in this area and we want your help in making improvements, more details here: #2238.
- Submit bug fixes, documentation changes, whatever. There's a never ending amount of work to be done around here and we need outside contributions to drive the project forward. Take a look at the issue tracker (in particular look for bugs or issues that are marked with the accepted label), talk with us on Python Discord, and read our (currently being overhauled) [contributing documentation](https://black.readthedocs.io/en/latest/contributing/index.html). We will be looking toward your patches!

---

One final note, I have been mentioned a lot here in terms of work done, but please know everyone else in the core team played and continues to play a critical role in this project. If it wasn't for their trust, knowledge, and dedication, I wouldn't be here. 

It has been a wild ride ever since I joined a year ago. Thank you for sticking by!

**Comment by agalazis** [comment]
Well  I am not willing to allow pre-release packages appart  from black `pipenv` 's configuration is giving me a headache see:
https://github.com/pypa/pipenv/issues/1760
How about denoting that it's not v1 yet via semver? ie have 0.something  releases until we reach to official release?

**Comment by astrojuanlu** [comment]
Going back to `0.*` is not possible, since present releases are `21.*`.

**Comment by agalazis** [comment]
not if the porject gets renamed (to black2 for example) that way we can have a new set of packages for black2 0.1.0 etc

**Comment by ruohola** [comment]
> not if the project gets renamed (to black2 for example) that way we can have a new set of packages for black2 0.1.0 etc

No thank you.

**Comment by layday** [comment]
It is possible to do using a version epoch, see https://www.python.org/dev/peps/pep-0440/#version-epochs.  (Not that I think it should be done here.)

**Comment by agalazis** [comment]
Acutally you can disregard my suggestion. All I wanted is not allowing pre-releases with pipenv but I didn't release this comment :
https://github.com/pypa/pipenv/issues/1760#issuecomment-677702984

Maybe we could add official instructions in readme/docs to use specific versions in pipenv  rather than *

**Comment by lig** [comment]
@WimYedema right. That is exactly what is happening;) Black is already the standard format for Python code in an enormous amount of projects. Black is more stable and usable than some libraries in Python's stdlib. I even changed a laptop and restored my Black sticker on a new one already. And still, we have it in a "pre-release" stage.
I'm starting to think that Black will be a legendary Python package that will never go out of a pre-release just for the sake of it.

**Comment by gsemet** [comment]
Hi. I have read and reread again this thread and I still do not understand why a non-prerelease version of black has been released.
At least provide the option to the user. Keep "black" as it is but provide a "black-stable" package that allow your beloved users to not force version or allow_prerelease for package. 

I use poetry, so this is not a problem for me. But I do not see why black would behave differently than any other python package (release... Pull Request... release.... and over and over).

**Comment by dimaqq** [comment]
@gsemet please release it to PYPI. call it `gray` or whatever 🖤

**Comment by hugovk** [comment]
https://pypi.org/project/gray/, https://pypi.org/project/grey/, https://pypi.org/project/blue/, https://pypi.org/project/tan/ and https://pypi.org/project/white/ already exist :)

**Comment by dimaqq** [comment]
Someone even made https://pypi.org/project/blacker/ which is black at some old version 😅

`blackish` is still free though!

**Comment by felix-hilden** [comment]
We appreciate the enthusiasm, but please note that we have defined a stability policy (#2394) and are planning a stable release. Any news will be announced here. In the meantime though, further comments don't really make the cogs turn any faster.

**Comment by dyve** [comment]
> We appreciate the enthusiasm, but please note that we have defined a stability policy (#2394) and are planning a stable release. Any news will be announced here. In the meantime though, further comments don't really make the cogs turn any faster.

Agreed. No black humor please.

**Comment by gsemet** [comment]
> https://pypi.org/project/blacker/

That's an option:
- get a pipeline triggered on any new version of package `black`
- convert the version `XX.M.bP` to `XX.0M.P` and rebuild the same package named `black-released`

**Comment by felix-hilden** [comment]
I'm pleased to announce that Black is finally non-beta software! 🎉 Here's the full [changelog](https://github.com/psf/black/releases/tag/22.1.0)!

Going forward we'll follow our [stability policy](https://black.readthedocs.io/en/stable/the_black_code_style/index.html#stability-policy). Work continues as usual with bugfixes and enhancements, but style changes are now introduced under our new `--preview` CLI switch. This allows us to evolve Black's style without too much disruption to users that want consistency. The default style is updated yearly.

Thanks to our maintainers for orchestrating the efforts, especially to our most recent reinforcement Batuhan (@isidentical) who was responsible for our match statement support! A hearty thank you to all of our contributors for pushing Black forward, and to our users for being the reason we do it!

Thanks for sticking with us, it's time to close this issue!

**Comment by ambv** [comment]
![im-not-crying](https://user-images.githubusercontent.com/55281/151678085-05911ef0-6450-4d66-965f-8edfdfbc1040.gif)
